transports/tcp/src/lib.rs

```diff
@@ -188,106 +183,6 @@ impl Config {
-    ///
-    /// [`Transport`] keeps track of the listen socket addresses as they
-    /// are reported by polling it. It is possible to listen on multiple
-    /// addresses, enabling port reuse for each, knowing exactly which listen
-    /// address is reused when dialing with a specific [`Transport`], as in the
-    /// following example:
-    ///
-    /// ```no_run
-    /// # use futures::StreamExt;
-    /// # use libp2p_core::transport::{ListenerId, TransportEvent};
-    /// # use libp2p_core::{Multiaddr, Transport};
-    /// # use std::pin::Pin;
-    /// # #[cfg(not(feature = "async-io"))]
-    /// # fn main() {}
-    /// #
-    /// #[cfg(feature = "async-io")]
-    /// #[async_std::main]
-    /// async fn main() -> std::io::Result<()> {
-    ///
-    /// let listen_addr1: Multiaddr = "/ip4/127.0.0.1/tcp/9001".parse().unwrap();
-    /// let listen_addr2: Multiaddr = "/ip4/127.0.0.1/tcp/9002".parse().unwrap();
-    ///
-    /// let mut tcp1 = libp2p_tcp::async_io::Transport::new(libp2p_tcp::Config::new().port_reuse(true)).boxed();
-    /// tcp1.listen_on(ListenerId::next(), listen_addr1.clone()).expect("listener");
-    /// match tcp1.select_next_some().await {
-    ///     TransportEvent::NewAddress { listen_addr, .. } => {
-    ///         println!("Listening on {:?}", listen_addr);
-    ///         let mut stream = tcp1.dial(listen_addr2.clone()).unwrap().await?;
-    ///         // `stream` has `listen_addr1` as its local socket address.
-    ///     }
-    ///     _ => {}
-    /// }
-    ///
-    /// let mut tcp2 = libp2p_tcp::async_io::Transport::new(libp2p_tcp::Config::new().port_reuse(true)).boxed();
-    /// tcp2.listen_on(ListenerId::next(), listen_addr2).expect("listener");
-    /// match tcp2.select_next_some().await {
-    ///     TransportEvent::NewAddress { listen_addr, .. } => {
-    ///         println!("Listening on {:?}", listen_addr);
-    ///         let mut socket = tcp2.dial(listen_addr1).unwrap().await?;
-    ///         // `stream` has `listen_addr2` as its local socket address.
-    ///     }
-    ///     _ => {}
-    /// }
-    /// Ok(())
-    /// }
-    /// ```
-    ///
-    /// If a wildcard listen socket address is used to listen on any interface,
-    /// there can be multiple such addresses registered for port reuse. In this
-    /// case, one is chosen whose IP protocol version and loopback status is the
-    /// same as that of the remote address. Consequently, for maximum control of
-    /// the local listening addresses and ports that are used for outgoing
-    /// connections, a new [`Transport`] should be created for each listening
-    /// socket, avoiding the use of wildcard addresses which bind a socket to
-    /// all network interfaces.
-    ///
-    /// When this option is enabled on a unix system, the socket
-    /// option `SO_REUSEPORT` is set, if available, to permit
-    /// reuse of listening ports for multiple sockets.
-    pub fn port_reuse(mut self, port_reuse: bool) -> Self {
```

Comment:
For many users, this might be the only breaking change they encounter if - for example - they don't implement any `NetworkBehaviour`s or `Transport`s themselves.

What do you think of - instead of removing it - marking it a `#[deprecated]` with a note of saying "does nothing because we decide on a per-connection basis now"? Currently, users would be left with a "this function no longer exists but I still want to enable port-reuse"-experience and they have to go and hunt down the PR that made this change, only to learn that it is actually done automatically for them now.

Reply:
I see your point I will mark it as deprecated. It could also offer a transition mode where, if you set the option we will always reuse ports.

Reply:
Yeah. If you fancy, you can explain in the doc comment how it works now :)